### PR TITLE
CSV Rejects Tables 2.0

### PR DESCRIPTION
**Pull request by pdet**
This PR extends the current implementation of the CSV Rejects Tables.

The following errors can now be registered in a rejects table:
1.  CAST
2. MISSING COLUMNS
3. TOO MANY COLUMNS
4. UNQUOTED VALUE
5. LINE SIZE OVER MAXIMUM
6. INVALID UNICODE

They will also produce two temporary tables.
**reject_scans**  - Stores information about the CSV Scan, including configuration.
Columns:
1. scan_id
2. file_id
3. delimiter
4. quote
5. escape
6. newline_delimiter
7. skip_rows
8. has_header
9. columns
10. date_format
11. timestamp_format
12. user_arguments

**reject_errors** - Stores errors that happened in a given scan.
Columns:
1. scan_id
2. file_id
3. line : Line number where error happened
4. line_byte_position: Byte position where faulty line starts
5. byte_positon: Byte position of the actual error
6. column_idx
7. column_name
8. error_type: Enum with error type
9. csv_line: Original CSV line where error happened
10. error_message: Error Message constructed in DuckDB.

There are also 4 parameters that can be used for the rejects tables
1. store_rejects: boolean, defines if we will store rejects, by default it stores on tables named **reject_scans** and **reject_errors** 
2. rejects_table: VARCHAR, optional table where errors will be stored. By default it's  **reject_errors**.
3. rejects_scans: VARCHAR, optional table where scan information about the errors will be stored. By default it's  **reject_scans**.
4. rejects_limit: BIGINT, optional variable indicating how many errors should be stored

Example:
```csv
name,age,current_day, barks
oogie boogie,3, 2023-01-01, 2
oogie boogie,3, 2023-01-02, 5
oogie boogie,3, 2023-01-03, bla, 7
oogie boogie,3, bla, bla, 7
"oogie boogie"bla,3, 2023-01-04
oogie boogie,3, bla
oogie boogieoogie boogieoogie boogieoogie boogieoogie boogieoogie boogieoogie boogie,3, bla
```

```sql
query IIII
FROM read_csv('data/csv/rejects/multiple_errors/multiple_errors.csv',
    columns = {'name': 'VARCHAR', 'age': 'INTEGER', 'current_day': 'DATE', 'barks': 'INTEGER'},
    store_rejects = true, auto_detect=false, header = 1, max_line_size=40);
----
oogie boogie	3	2023-01-01	2
oogie boogie	3	2023-01-02	5

query IIIIIIIIIIIII rowsort
FROM reject_scans ORDER BY ALL;
----
71	0	data/csv/rejects/multiple_errors/multiple_errors.csv	,	"	"	\n	0	true	{'name': 'VARCHAR','age': 'INTEGER','current_day': 'DATE','barks': 'INTEGER'}	NULL	NULL	max_line_size='40', header=true, store_rejects=true


query IIIIIIIIII rowsort
FROM reject_errors ORDER BY ALL;
----
71	0	4	89	116	4	barks	CAST	oogie boogie,3, 2023-01-03, bla, 7	Error when converting column "barks". Could not convert string " bla" to 'INTEGER'
71	0	4	89	120	5	NULL	TOO MANY COLUMNS	oogie boogie,3, 2023-01-03, bla, 7	Expected Number of Columns: 4 Found: 5
71	0	5	124	144	4	barks	CAST	oogie boogie,3, bla, bla, 7	Error when converting column "barks". Could not convert string " bla" to 'INTEGER'
71	0	5	124	148	5	NULL	TOO MANY COLUMNS	oogie boogie,3, bla, bla, 7	Expected Number of Columns: 4 Found: 5
71	0	6	152	152	1	name	UNQUOTED VALUE	"oogie boogie"bla,3, 2023-01-04	Value with unterminated quote found.
71	0	6	152	183	3	barks	MISSING COLUMNS	"oogie boogie"bla,3, 2023-01-04	Expected Number of Columns: 4 Found: 3
71	0	7	184	203	3	barks	MISSING COLUMNS	oogie boogie,3, bla	Expected Number of Columns: 4 Found: 3
71	0	8	204	204	NULL	NULL	LINE SIZE OVER MAXIMUM	oogie boogieoogie boogieoogie boogieoogie boogieoogie boogieoogie boogieoogie boogie,3, bla	Maximum line size of 40 bytes exceeded. Actual Size:92 bytes.
71	0	8	204	295	3	barks	MISSING COLUMNS	oogie boogieoogie boogieoogie boogieoogie boogieoogie boogieoogie boogieoogie boogie,3, bla	Expected Number of Columns: 4 Found: 3

```

Also, if multiple errors occur on the same line, they will all be reported. The exception to this is related to casting errors during the Flush method (that is, not implicit casts). This situation should improve as we implement more implicit casts.

### CODE REVIEW

**Comment by darthf1** [comment]
This is awesome!

Would it be possible to add something like MISSING_VALUES? Where a column which does exist, but with a NULL value, will error out?

I'm currently doing CSV validation via [JSON schema](https://python-jsonschema.readthedocs.io/en/stable), where I check datatypes, missing columns, and missing values. 

Reading this PR description I can replace it (almost fully) with the rejects table in duckdb which will be a lot more performant.

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -773,6 +773,17 @@ CatalogEntry &Catalog::GetEntry(ClientContext &context, const string &schema, co
 	throw CatalogException("CatalogElement \"%s.%s\" does not exist!", schema, name);
 }
 
+bool Catalog::EntryExists(ClientContext &context, const string &schema, const string &name) {
```
Comment (Mytherin) [inline src/catalog/catalog.cpp:776]: Can this be removed and can we use ` GetEntry` instead?

File: third_party/utf8proc/utf8proc_wrapper.cpp [hunk 2]
```diff
@@ -102,6 +102,39 @@ UnicodeType Utf8Proc::Analyze(const char *s, size_t len, UnicodeInvalidReason *i
...
+			/* invalid UTF-8 start byte */
+			s[i] = special_flag; // Rewrite invalid byte
+		}
+		if (type == UnicodeType::INVALID) {
+			for (size_t j = first_pos_seq; j <= i; j++) {
+                s[j] = special_flag; // Rewrite each byte of the invalid sequence
+            }
+			type = UnicodeType::ASCII;
+		}
+	}
```
Comment (Mytherin) [inline third_party/utf8proc/utf8proc_wrapper.cpp:135]: Maybe we can add a `D_ASSERT(Utf8::IsValid())` after the cleanup?

File: third_party/utf8proc/utf8proc_wrapper.cpp [hunk 3]
```diff
@@ -102,6 +102,39 @@ UnicodeType Utf8Proc::Analyze(const char *s, size_t len, UnicodeInvalidReason *i
 	return type;
 }
 
+void Utf8Proc::MakeValid(char *s, size_t len, char special_flag){
+	UnicodeType type = UnicodeType::ASCII;
```
Comment (Mytherin) [inline third_party/utf8proc/utf8proc_wrapper.cpp:106]: Maybe assert that `special_flag` is `<= 127`

File: test/sql/copy/csv/rejects/csv_rejects_two_tables.test [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+0	4176	20876	20876	1	column0	CAST	C, A	Error when converting column "column0". Could not convert string "C" to 'BIGINT'
+1	3680	18396	18396	1	column0	CAST	B, A	Error when converting column "column0". Could not convert string "B" to 'BIGINT'
+1	5680	28396	28396	1	column0	CAST	C, A	Error when converting column "column0". Could not convert string "C" to 'BIGINT'
+
+# Test giving the name of both tables
+query IIIII
+SELECT typeof(first(column0)), typeof(first(column1)), COUNT(*), SUM(column0), MAX(len(column1)) FROM read_csv_auto(
+    'test/sql/copy/csv/data/error/mismatch/big_bad*.csv',
+    sample_size=1,
+    rejects_scan = 'rejects_scan_3',
```
Comment (Mytherin) [inline test/sql/copy/csv/rejects/csv_rejects_two_tables.test:99]: Could you add a test that checks what happens if both `rejects_scan` and `rejects_table` are the same name (but the table does not exist)?

File: src/storage/serialization/serialize_nodes.cpp [hunk 5]
```diff
@@ -135,26 +135,26 @@ void CSVReaderOptions::Serialize(Serializer &serializer) const {
...
+	serializer.WriteProperty<CSVOption<char>>(121, "dialect_options.state_machine_options.escape", dialect_options.state_machine_options.escape);
+	serializer.WriteProperty<CSVOption<bool>>(122, "dialect_options.header", dialect_options.header);
+	serializer.WritePropertyWithDefault<idx_t>(123, "dialect_options.num_cols", dialect_options.num_cols);
+	serializer.WriteProperty<CSVOption<NewLineIdentifier>>(124, "dialect_options.state_machine_options.new_line", dialect_options.state_machine_options.new_line);
+	serializer.WriteProperty<CSVOption<idx_t>>(125, "dialect_options.skip_rows", dialect_options.skip_rows);
+	serializer.WriteProperty<map<LogicalTypeId, CSVOption<StrpTimeFormat>>>(126, "dialect_options.date_format", dialect_options.date_format);
+	serializer.WritePropertyWithDefault<string>(127, "sniffer_user_mismatch_error", sniffer_user_mismatch_error);
+	serializer.WritePropertyWithDefault<bool>(128, "parallel", parallel);
+	serializer.WriteProperty<CSVOption<string>>(129, "rejects_table_name", rejects_table_name);
+	serializer.WriteProperty<CSVOption<string>>(130, "rejects_scan_name", rejects_scan_name);
```
Comment (Mytherin) [inline src/storage/serialization/serialize_nodes.cpp:151]: This should probably use the `WritePropertyWithDefault` (i.e. we need to specify a default value for `CSVOption`)

File: src/storage/serialization/serialize_nodes.cpp [hunk 6]
```diff
@@ -135,26 +135,26 @@ void CSVReaderOptions::Serialize(Serializer &serializer) const {
 	serializer.WritePropertyWithDefault<idx_t>(114, "buffer_size", buffer_size);
 	serializer.WriteProperty<MultiFileReaderOptions>(115, "file_options", file_options);
 	serializer.WritePropertyWithDefault<vector<bool>>(116, "force_quote", force_quote);
-	serializer.WritePropertyWithDefault<string>(117, "rejects_table_name", rejects_table_name);
+	serializer.WriteProperty<CSVOption<bool>>(117, "store_rejects", store_rejects);
```
Comment (Mytherin) [inline src/storage/serialization/serialize_nodes.cpp:138]: Could we try to minimize the breakage, and use default properties where possible?

File: src/include/duckdb/transaction/transaction_context.hpp [hunk 7]
```diff
@@ -48,6 +48,7 @@ class TransactionContext {
 	}
 
 	idx_t GetActiveQuery();
+	idx_t GetIncrementalIndex();
```
Comment (Mytherin) [inline src/include/duckdb/transaction/transaction_context.hpp:51]: Can we move this to the `ObjectCache` instead? you could e.g. do this per query number, and then reset when a new query number is encountered

File: src/include/duckdb/catalog/catalog.hpp [hunk 8]
```diff
@@ -232,6 +232,9 @@ class Catalog {
 	//! Gets the "schema.name" entry without a specified type, if entry does not exist an exception is thrown
 	DUCKDB_API CatalogEntry &GetEntry(ClientContext &context, const string &schema, const string &name);
 
+	//! Returns true if the "schema.name" entry without a specified type exists
+	DUCKDB_API bool EntryExists(ClientContext &context, const string &schema, const string &name);
```
Comment (Mytherin) [inline src/include/duckdb/catalog/catalog.hpp:236]: Can this be removed?

File: src/include/duckdb/storage/serialization/nodes.json [hunk 9]
```diff
@@ -604,61 +604,61 @@
...
         "name": "rejects_limit",
         "type": "idx_t"
       },
       {"id": 119,
-        "name": "rejects_recovery_columns",
-        "type": "vector<string>"
-      },
-      {"id": 120,
-        "name": "rejects_recovery_column_ids",
-        "type": "vector<idx_t>"
```
Comment (Mytherin) [inline src/include/duckdb/storage/serialization/nodes.json:620]: "deleted": true to delete fields (you cannot delete & rename fields)

File: src/include/duckdb/storage/serialization/nodes.json [hunk 10]
```diff
@@ -604,61 +604,61 @@
         "type": "vector<bool>"
       },
       {"id": 117,
-        "name": "rejects_table_name",
-        "type": "string"
+        "name": "store_rejects",
+        "type": "CSVOption<bool>"
```
Comment (Mytherin) [inline src/include/duckdb/storage/serialization/nodes.json:608]: We can't change the type of fields, unless the serialization of the types is identical

File: src/include/duckdb/storage/serialization/nodes.json [hunk 11]
```diff
@@ -537,7 +537,7 @@
     "members": [
       {"id": 100,
         "name": "ignore_errors",
-        "type": "bool"
+        "type": "CSVOption<bool>"
```
Comment (Mytherin) [inline src/include/duckdb/storage/serialization/nodes.json:540]: Is the serialization of `bool` identical to `CSVOption<bool>`? If not we can't change this in this manner. Could we just serialize this as a boolean? e.g. using `"serialize_property"`

File: src/include/duckdb/storage/serialization/nodes.json [hunk 12]
```diff
@@ -604,61 +604,61 @@
...
         "type": "string"
       },
-      {"id": 130,
+      {"id": 128,
         "name": "parallel",
         "type": "bool"
       },
+      {"id": 129,
+        "name": "rejects_table_name",
+        "type": "CSVOption<string>"
```
Comment (Mytherin) [inline src/include/duckdb/storage/serialization/nodes.json:656]: Can we add a default value?

File: src/execution/operator/csv_scanner/util/csv_error.cpp [hunk 13]
```diff
@@ -102,40 +119,53 @@ CSVError CSVError::ColumnTypesError(case_insensitive_map_t<idx_t> sql_types_per_
...
-		      << '\n';
+		how_to_fix_it
+		    << "This type was either manually set or derived from an existing table. Select a different type to "
+		       "correctly parse this column."
+		    << '\n';
 	}
-	error << options.ToString();
 
-	return CSVError(error.str(), CSVErrorType::CAST_ERROR, column_idx, row, error_info);
+	return CSVError(error.str(), CSVErrorType::CAST_ERROR, column_idx, csv_row, error_info, row_byte_position,
```
Comment (Mytherin) [inline src/execution/operator/csv_scanner/util/csv_error.cpp:153]: Could we use the new code to generate the original CSV line here as well?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments below:

**Comment by pdet** [comment]
> This is awesome!
> 
> Would it be possible to add something like MISSING VALUES? Where a column which does exist, but with a NULL value, will error out?
> 
> I'm currently doing CSV validation via [JSON schema](https://python-jsonschema.readthedocs.io/en/stable), where I check datatypes, missing columns, and missing values.
> 
> Reading this PR description I can replace it almost fully, in my usecase at least, with the rejects table in duckdb which will be a lot more performant.
> 
> But I recognize that NULL values have nothing to do with a valid / invalid CSV structure :)

Absolutely. I think that's a bit orthogonal to this PR, but I'm happy to add something in this direction in the near-future. :-)

**Comment by Mytherin** [comment]
Thanks!

**Comment by adriens** [comment]
:clap: Awesome :star_struck:

**Comment by aborruso** [comment]
@pdet first of all thank you very much, it's really useful, it's really a great job.

I don't understand one of the errors in your example, `SIZE OVER MAXIMUM`: "Maximum line size of 40 bytes exceeded. Actual Size:92 bytes".

Is there a limit? And where is this limit set?

Thank you again

**Comment by aborruso** [comment]
> Is there a limit? And where is this limit set?

I'm stupid :)

We have `max_line_size=40`

**Comment by aborruso** [comment]
Hi @pdet I have tested and I have some notes.

My input file:

```
nome,compleanno,altezza
Maurizio,1992-12-27,187
Paola,gennaio,162
Andy,1973-07-06,176,Palermo

Chiara,1991-02-02,162
```

If I run

```
FROM read_csv('tmp.csv',columns = {'nome': 'VARCHAR', 'compleanno': 'DATE', 'altezza': 'INTEGER'}, store_rejects = true, auto_detect=false, header = 1);
```

I have the below errors. My notes:

- In the error log I read a CAST in line `5`. It doesn't seem right to me, I have a cast error in line `3`;
- reading the file, it correctly skips line 5, but I cannot find the error in the log. How about introducing the "empty row" error?

| scan_id | file_id | line | line_byte_position | byte_position | column_idx | column_name | error_type | csv_line | error_message |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 7 | 0 | 4 | 67 | 86 | 4 |  | TOO MANY COLUMNS | Andy,1973-07-06,176,Palermo | Expected Number of Columns: 3 Found: 4 |
| 7 | 0 | 5 | 49 |  | 2 | compleanno | CAST | Paola,gennaio,162 | Error when converting column "compleanno". date field value out of range: "gennaio", expected format is (YYYY-MM-DD) |

**Comment by aborruso** [comment]
@pdet I'm adding another note.

If I have this

```
nome,compleanno,altezza
Maurizio,1992-12-27,187
Paola,gennaio,162
Andy,1973-07-06,176,Palermo

Chiara,02-02,162
Mario
```

I have a cast error in row 7: "Error when converting column ""compleanno"". date field value out of range: ""gennaio"", expected format is (YYYY-MM-DD)"
That cast error is in row 6.

**Pull request merged**